File: dart/lib/src/utils/deserialize_reverse.dart
import 'dart:convert';
import 'dart:typed_data';

/// Parameters for deserializing reverse account data
class DeserializeReverseParams {
  const DeserializeReverseParams({
    required this.data,
    this.trimFirstNullByte = false,
  });

  /// The data to deserialize
  final Uint8List? data;

  /// Whether to trim the first null byte from the result string
  final bool trimFirstNullByte;
}

/// Deserializes reverse account data.
///
/// This function mirrors js-kit/src/utils/deserializers/deserializeReverse.ts
///
/// [data] - The Uint8List to deserialize. If null, returns null.
/// [trimFirstNullByte] - Whether to trim the first null byte from the result string
///
/// Returns the deserialized string, or null if data is null.
String? deserializeReverse(Uint8List? data, {bool trimFirstNullByte = false}) {
  if (data == null) return null;

  final byteData = ByteData.sublistView(data);
  final nameLength = byteData.getUint32(0, Endian.little);

  if (data.length < 4 + nameLength) return null;

  final nameBytes = data.sublist(4, 4 + nameLength);
  var result = utf8.decode(nameBytes);

  if (trimFirstNullByte && result.startsWith('\u0000')) {
    result = result.substring(1);
  }

  return result;
}

File: js-kit/src/utils/deserializers/deserializeReverse.ts
import { ReadonlyUint8Array } from "@solana/kit";

import { utf8Codec } from "../../codecs";

/**
 * Parameters for deserializing reverse account data.
 *
 * @example
 * ```ts
 * const params: DeserializeReverseParams = { data: reverseAccountData };
 * ```
 */
export interface DeserializeReverseParams {
  /** Reverse account data. */
  data: ReadonlyUint8Array | undefined;
  /** Whether to remove a subdomain's leading null byte. Defaults to false. */
  trimFirstNullByte?: boolean;
}

/**
 * Deserializes reverse account data.
 *
 * The first four bytes encode the reverse name length.
 *
 * @param params Reverse deserialization parameters
 * @param params.data Reverse account data. If undefined, returns undefined
 * @param params.trimFirstNullByte Whether to trim the first null byte for subdomain reverse names. Defaults to false
 * @returns The deserialized string, or `undefined` if data is undefined.
 *
 * @example
 * ```ts
 * const name = deserializeReverse({ data: reverseAccountData });
 * ```
 */
export function deserializeReverse({
  data,
  trimFirstNullByte,
}: DeserializeReverseParams): string;

export function deserializeReverse({
  data,
  trimFirstNullByte,
}: DeserializeReverseParams): undefined;

export function deserializeReverse({
  data,
  trimFirstNullByte = false,
}: DeserializeReverseParams): string | undefined {
  if (!data) return undefined;

  const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
  const nameLength = view.getUint32(0, true);

  return utf8Codec
    .decode(data.subarray(4, 4 + nameLength))
    .replace(/^\0/, trimFirstNullByte ? "" : "\0");
}
